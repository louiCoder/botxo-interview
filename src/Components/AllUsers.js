import React from 'react';
import { useDispatch } from 'react-redux';
import { removeGenericUserAction, removeSpecificUserAction } from '../Store/ActionCreators';

export default function AllUsers ({ users, header, genericOrSpecific }) {
  const dispatch = useDispatch();
  const confirmString = (name) => `Are you sure you want to delete ${name} from the`;

  // remove user from either specific or generic list
  const removeUserHandler = (name) => {
    switch (genericOrSpecific) {
      case 'generic':
        const confirmDelGeneral = window.confirm(`${confirmString(name)} general list`);
        if (!confirmDelGeneral) return;
        return dispatch(removeGenericUserAction(name));
      case 'specific':
        const confirmDelSpecific = window.confirm(`${confirmString(name)} pecific list`);
        if (!confirmDelSpecific) return;
        return dispatch(removeSpecificUserAction(name));
      default:
        return;
    }
  };
  return (
    <div>
      <h3>{header}</h3>
      <div className="all-users-list-container">
        {users && (
          <ul style={{}}>
            {users.map(({ name, age, fromGeneric }, index) => (
              <div key={index} className={`single-name ${fromGeneric ? 'disabled-name' : ''}`}>
                <li disabled={fromGeneric ? true : false}>
                  {index + 1}. {name} ({age})
                  <span onClick={() => (fromGeneric ? null : removeUserHandler(name))}>
                    {' '}
                    <svg style={{ width: '24px', height: '24px' }} viewBox="0 0 24 24">
                      <path
                        fill="currentColor"
                        d="M19,4H15.5L14.5,3H9.5L8.5,4H5V6H19M6,19A2,2 0 0,0 8,21H16A2,2 0 0,0 18,19V7H6V19Z"
                      />
                    </svg>
                  </span>
                </li>
              </div>
            ))}
          </ul>
        )}
      </div>
    </div>
  );
}
