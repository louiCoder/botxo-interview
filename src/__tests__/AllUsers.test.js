import React from 'react';
import { mount } from 'enzyme';
import Fixtures from '../Fixture';
import mockStore from '../Store/MockStore';
import { Provider } from 'react-redux';
import AllUsers from '../Components/AllUsers';

describe('<AllUsers /> component', () => {
  let wrapper, USERS, store, props;
  beforeEach(() => {
    store = mockStore({ genericUsers: [], specificUsers: [], loading: false });
    USERS = Fixtures;
    props = {
      users: USERS,
      header: 'generic',
      genericOrSpecific: 'generic',
      removeUserHandler: jest.fn()
    };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );
  });

  afterEach(() => {
    USERS = Fixtures;
  });

  it('Render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render h3 header with Generic word when header="generic" ', () => {
    expect(wrapper.find('h3').props().children).toEqual('generic');
  });

  it('should render 32 list items', () => {
    expect(wrapper.find('li').length).toEqual(32);
  });

  it('should call confirm function when user confirms action in the dialog', () => {
    window.confirm = jest.fn(() => true);
    wrapper.find('li').first().find('span').simulate('click', 'Alberto');
    expect(window.confirm).toHaveBeenCalled();
  });

  it('should not dispatch REMOVE_GENERIC_USER_ACTION when user cancels dialog', () => {
    window.confirm = jest.fn(() => false);
    wrapper.find('li').first().find('span').simulate('click', 'Alberto');
    expect(store.getActions()).toEqual([]);
  });

  it('should not dispatch REMOVE_SPECIFIC_USER_ACTION when user cancels dialog', () => {
    props = { ...props, genericOrSpecific: 'specific', header: 'specific' };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );
    window.confirm = jest.fn(() => false);
    wrapper.find('li').first().find('span').simulate('click', 'Alberto');
    expect(store.getActions()).toEqual([]);
  });

  it('should render first item as disabled item when genericOrSpecific property is true', () => {
    const _USERS = (USERS[0] = { ...Fixtures[0], fromGeneric: true });
    props = { ...props, USERS: _USERS };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );
    expect(wrapper.find('li').first().prop('disabled')).toBeTruthy();
  });

  it('should dispatch REMOVE_GENERIC_USER_ACTION when element delete icon is clicked and user confirms', () => {
    const _USERS = (USERS[0] = { ...Fixtures[0], fromGeneric: false });
    props = { ...props, USERS: _USERS };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );

    window.confirm = jest.fn(() => true);
    wrapper.find('li').first().find('span').simulate('click', 'Alberto');
    expect(store.getActions()).toEqual([ { type: 'REMOVE_GENERIC_USER_ACTION', name: 'Alberto' } ]);
  });

  it('should dispatch REMOVE_SPECIFIC_USER_ACTION when element delete icon is clicked and user confirms', () => {
    props = { ...props, genericOrSpecific: 'specific', header: 'specific' };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );
    window.confirm = jest.fn(() => true);
    wrapper.find('li').last().find('span').simulate('click', 'Santiago');
    expect(store.getActions()).toEqual([ { type: 'REMOVE_SPECIFIC_USER_ACTION', name: 'Santiago' } ]);
  });

  it('should fall back to default case when genericOrSpecific is undefined', () => {
    props = { ...props, genericOrSpecific: undefined };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );
  });

  it('should not execute removeUserHandler function when fronGeneric is false', () => {
    USERS[0] = { ...USERS[0], fromGeneric: true };
    props = { ...props, USERS, genericOrSpecific: 'specific', header: 'specific' };
    wrapper = mount(
      <Provider store={store}>
        <AllUsers {...props} />
      </Provider>
    );
    wrapper.find('li').first().find('span').simulate('click', 'Santiago');
    expect(store.getActions()).toEqual([]);
  });
});
