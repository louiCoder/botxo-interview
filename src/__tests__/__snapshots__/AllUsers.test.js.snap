// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AllUsers /> component Render without crashing 1`] = `ReactWrapper {}`;
